compiler_gym/util/minimize_trajectory.py

```diff
@@ -0,0 +1,317 @@
+            mid,
+            len(all_actions),
+            env.action_space.flags[all_actions[mid]],
+        )
+
+    if not apply_and_test(actions):
+        raise MinimizationError("Post-bisect sanity check failed!")
+
+    yield env
+
+
+def random_minimization(
+    env: CompilerEnv,
+    hypothesis: Hypothesis = environment_validation_fails,
+    num_steps_ratio_multiplier: float = 5,
+    init_discard_ratio: float = 0.75,
+    discard_ratio_decay: float = 0.75,
+    flakiness: int = 1,
+) -> Iterable[CompilerEnv]:
+    """Run an iterative process of randomly removing actions to minimize a
+    trajectory.
+
+    For each round of minimization, a number of actions are discarded randomly
+    and the hypothesis is tested. If the hypothesis still holds with those
+    actions removed, the minimization proceeds. Else the actions are re-inserted
+    into the trajectory and a new set of actions are removed. After a failure
+
+    Performs up to O(num_steps_ratio_multiplier * log n) evaluation rounds,
+    where n is the length of the trajectory.
+
+    :param env: An environment whose action trajectory should be minimized.
+    :param hypothesis: The hypothesis that is used to determine if a trajectory
+        is valid. A callback that accepts as argument the :code:`env:`
+        instance and returns true if the hypothesis holds, else false. The
+        hypothesis must hold on the initial trajectory.
+    :param num_steps_ratio_multiplier: A multiplier for the number of rounds of
+        minimization to perform, using log(n) the length of the trajectory as
+        the factor.
+    :param init_discard_ratio: The number of actions that will be randomly
+        discarded, as a multiplier of the length of the trajectory.
+    :param discard_ratio_decay: The ratio of decay for the discard ratio on
+        failure.
+    :param flakiness: The maximum number of times the hypothesis is repeated
+        to check if it holds.
+    :returns: A generator that yields the input environment every time the
+        trajectory is successfully reduced.
+    :raises MinimizationError: If the minimization fails, or if the hypothesis
+        does not hold on the initial trajectory.
+    """
+
+    def apply_and_test(actions):
+        return _apply_and_test(env, actions, hypothesis, flakiness)
+
+    actions = env.actions.copy()
+    if not apply_and_test(actions):
+        raise ValueError(
+            "Hypothesis failed on the initial state! The hypothesis must hold for the first state."
+        )
+
+    max_num_steps = int(log(len(actions), 2) * num_steps_ratio_multiplier)
+
+    num_steps = 0
+    discard_ratio = init_discard_ratio
+    while len(actions) > 5 and num_steps < max_num_steps:
+        num_steps += 1
+        num_to_remove = int(ceil(len(actions) * discard_ratio))
+        candidate_actions = actions.copy()
+        for _ in range(num_to_remove):
+            candidate_actions.remove(random.choice(candidate_actions))
```

Comment:
Not 100% sure how the underlying code works here. Is this removing by index or value?
If removing by value, then you might be removing all `reg2mem` for example. If there are repeated actions then that might not be what is expected.

Reply:
Ah, great catch! `remove()` is removing by value. So this means if there are duplicates, only the first instance is removed. I don't want to do this, so I'll replace it with an index-based approach.